### DISCUSSION TITLE
Use "real" python code inside of jit

### DISCUSSION THREAD

**Discussion by inversecrime** [jax-ml/jax#18130] [Q&A] [answered]
My functions often compute values that do not depend on the contents of any of the inputs, but rather on the shape of the inputs or not on the inputs at all.
This means that theoretically, it would be possible to precompute these values when jitting these functions, since jit assumes that the inputs have fixed shapes (as far as i understand, calling a jitted function with inputs of different shapes just triggers another compilation).

Is there any way to stage these precomputations out of jit?
A simple example would be the following code:

```
from functools import cached_property

import jax
import jax.numpy as jnp
from jax import Array


class Foo:
    @cached_property
    def some_property(self) -> Array:
        return jnp.ones((10, 10))


foo = Foo()
foo.some_property # depending on whether or not this is called, either (1) or (2) happens
jax.jit(lambda: foo.some_property)()
print(foo.some_property)  # (1) returns 10x10 ones (2) returns a traced array

```

This behavior is exactly what we expect from jit.
However, I want to achieve result (1) _without_ calling the property before the jit command.

**Comment by jakevdp** [comment, marked as answer]
It sounds like you're looking for [`jax.ensure_compile_time_eval`](https://jax.readthedocs.io/en/latest/_autosummary/jax.ensure_compile_time_eval.html):
```python
class Foo:
    @cached_property
    def some_property(self) -> Array:
        with jax.ensure_compile_time_eval():
            return jnp.ones((10, 10))
```

**Reply by inversecrime** [reply]
Thanks!

**Reply by inversecrime** [reply]
Would you recommend creating a custom property decorator for this exact behavior?
This could prevent unwanted results like case (2) in my example.

**Reply by jakevdp** [reply]
Maybe - but in general you should be careful about using JIT with side-effecting code (here the cacheing of the property is a side-effect). For more on this, see [JAX Sharp Bits: Pure Functions](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#pure-functions).

**Reply by inversecrime** [reply]
Usually all functions I use are pure, so this is not a problem.
I have another question though.
Consider the following case:

```
def g(x: Array) -> Array:  # this function's output only depends on the shape of x
    ...


def f(x: Array) -> Array:
    y = g(x)  # i want to precompute this, but with jit enabled
    return y * x
```

This seems not to work well:

```
def f(x: Array) -> Array:
    with jax.ensure_compile_time_eval():
        y = jax.jit(g)(x)
    return y * x
```

Another one:
I played around a bit with jit, jit.lower, jit.lower.compile and make_jaxpr.
In https://jax.readthedocs.io/en/latest/aot.html, it is mentioned that some values get simplified - there is an example where 2*7 = 14 gets precomputed.
However, it seems like that in general, compiling does not precompute variables which only depend on constants and input array shapes.
Is this true or wrong?

I am new to github, if my questions are in the wrong place, please let me know.

**Reply by jakevdp** [reply]
I wonder if this question is a form of premature optimization – under JIT, the XLA compiler will de-duplicate repeated operations, so perhaps the best approach is to not worry about cacheing at all, and just let the compiler do its job.

**Reply by inversecrime** [reply]
The thing is, I want to solve PDEs, and usually, creating the "pde sparse matrix" is computationally expensive (not as much as the solver of course, but still relevant).
When solving multiple PDEs with the same matrix, the compiler does not seem to recognize that the matrix does not change from call to call, and the jitted function recomputes it every single time.

Using vmap works well for this (then the matrix only gets created once), but sometimes it is not possible to use it.

I guess the obvious solution would be to create the matrix in a seperate function and then only jit the solver.

**Reply by jakevdp** [reply]
Yeah, that sounds better than trying to do caching via side-effects.

**Reply by inversecrime** [reply]
I don't really understand why caching is a side-effect. Would it be better practise to do this:
```
class Foo:
    def __init__(self) -> None:
        self.some_property = jnp.ones((10, 10))
```
Than this:
```
class Foo:
    @cached_property
    def some_property(self) -> Array:
        with jax.ensure_compile_time_eval():
            return jnp.ones((10, 10))
```
?

Is calling `some_property` considered a side-effect in both cases?
From my understanding after reading https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#pure-functions, it would not be a side effect if you pass an object of Class Foo to the corresponding function, like this:

```
@register_static
class Foo:
    def __init__(self) -> None:
        self.some_property = jnp.ones((10, 10))


def f(foo: Foo, x: Array) -> Array:
    return foo.some_property * x


print(jax.jit(f)(Foo(), jnp.ones(10)))
```

**Reply by jakevdp** [reply]
Cacheing is a side-effect because it mutates the object in-place without returning the mutated object: in other words, as a side-effect of calling the function, the object's contents change.

In the second case, it's also a side-effect, but it's a side-effect that's carefully constructed so as to not fall afoul of the assumptions made by JAX's transformation infrastructure. It's probably not a perfect solution, which is why I suggested a more direct approach.

**Comment by patrick-kidger** [comment]
To add on to the answer here. Compiling will generally precompute those values that depend only on array shapes and known constants. However, there are some exceptions to this. One example is that `a * 0` will not reduce to `0`, because `nan * 0 == nan`, so that's not a valid optimisation to make. This has quite far-reaching consequences; for example it's the reason that Jacobian computations often have an extra matmul against the identity matrix. (Which does *not* perform an identity operation if the function has NaNs.)

Another exception is if the number of dimensions in the constants get too large. Past some point the constant propagation will judge that it's too expensive, and defer the computation to runtime.

---

Wrt the specifics of your problem here, indeed explicitly precomputing the desired matrix in advance is probably what you want.

By the way, if you're solving differential equations, then you might like [Diffrax](https://github.com/patrick-kidger/diffrax). It's mostly targeting ODE/SDE solving, but it can handle [some semidiscretised PDEs](https://docs.kidger.site/diffrax/examples/nonlinear_heat_pde/) as well.

**Reply by inversecrime** [reply]
Thanks for your reply.
Indeed, I'm mainly interested in the wave equation and solving it with semidiscrete methods. I'll have a look at Diffrax.
I guess there is no scheme for 2nd order odes? Of course, I could just represent it as 1st order ode.
